core/src/main/java/net/adoptopenjdk/icedteaweb/resources/downloader/PrioritizedParallelExecutor.java

```diff
@@ -0,0 +1,218 @@
+     */
+    private static class CancelingCallable<V> implements Callable<V> {
+
+        private final AtomicBoolean complete = new AtomicBoolean(false);
+        private final AtomicReference<Future<V>> lowerPriority = new AtomicReference<>();
+
+        private final Callable<V> delegate;
+
+        private CancelingCallable(Callable<V> delegate) {
+            this.delegate = delegate;
+        }
+
+        /**
+         * Set a lower priority to be canceled.
+         * This is not passed in the constructor as the value is not known then.
+         *
+         * If this callable has already been completed successfully then the passed future will be canceled immediately.
+         *
+         * @param lowerPriority the lower priority future to be canceled after completing this callable successfully.
+         */
+        void setLowerPriority(Future<V> lowerPriority) {
+            this.lowerPriority.set(lowerPriority);
+            if (complete.get()) {
+                lowerPriority.cancel(true);
+            }
+        }
+
+        @Override
+        public V call() throws Exception {
+            final V result = delegate.call();
+            cancelLowerPriority();
+            return result;
+        }
+
+        private void cancelLowerPriority() {
+            complete.set(true);
+            final Future<V> lowerPriority = this.lowerPriority.get();
+            if (lowerPriority != null) {
+                lowerPriority.cancel(true);
+            }
+        }
+    }
+
+    /**
+     * Wrapper around {@link Future} which allows to construct a chain of futures.
+     * Lower priority futures are canceled if this future is canceled.
+     * If this future fails the result of the lower priority is returned.
+     */
+    private static class ChainableFuture<V> implements Future<V> {
+
+        private final AtomicBoolean terminatedExceptionally = new AtomicBoolean(false);
+        private final CompletableFuture<V> lowerPriorityResult = new CompletableFuture<>();
+        private final AtomicReference<Future<V>> lowerPriority = new AtomicReference<>();
+
+        private final Future<V> delegate;
+
+        private ChainableFuture(Future<V> delegate) {
+            this.delegate = delegate;
+        }
+
+        /**
+         * Set a lower priority to be canceled.
+         * This is not passed in the constructor as the value is not known then.
+         *
+         * @param lowerPriority the lower priority future to be canceled this future is canceled.
+         */
+        void setLowerPriority(Future<V> lowerPriority) {
+            this.lowerPriority.set(lowerPriority);
+
+            if (terminatedExceptionally.get()) {
+                // TODO: should the body of this if be executed in a separate thread to not block the current thread
+                //  which is about to plug together the future chain...
+                try {
+                    lowerPriorityResult.complete(lowerPriority.get());
```

Comment:
by doing this the method will block until the `lowerPrioroty` is done. I would not assume something like this in a setter method